File: apps/api/app/schemas/transactions.py
from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field


class AccountResponse(BaseModel):
    id: int
    name: str
    account_type: str
    institution: str | None = None
    current_balance: float
    cash_projection_role: str
    is_manual: bool
    plaid_account_id: str | None = None
    plaid_item_id: int | None = None
    mask: str | None = None

    model_config = ConfigDict(from_attributes=True)


class CategoryResponse(BaseModel):
    id: int
    name: str
    kind: str
    monthly_target: float
    is_default: bool
    budget_group_key: str | None = None
    budget_sort_order: int | None = None
    can_manage: bool | None = None

    model_config = ConfigDict(from_attributes=True)


class TransactionSplitResponse(BaseModel):
    id: int
    category: CategoryResponse
    amount: float
    notes: str | None = None

    model_config = ConfigDict(from_attributes=True)


class TransactionResponse(BaseModel):
    id: int
    posted_date: date
    description: str
    merchant: str | None = None
    amount: float
    transaction_type: str
    source_name: str | None = None
    import_hash: str
    notes: str | None = None
    plaid_transaction_id: str | None = None
    plaid_metadata: str | None = None
    pending: bool
    display_merchant: str
    raw_description: str | None = None
    plaid_category_label: str | None = None
    payment_channel_label: str | None = None
    location_summary: str | None = None
    category: CategoryResponse | None = None
    account: AccountResponse | None = None
    splits: list[TransactionSplitResponse] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)


class TransactionBudgetActionResponse(BaseModel):
    # Flask 964c369 transaction_budget_action payload: the "make this a
    # budget" hint shown next to categorized expense transactions.
    category_name: str
    target: float
    target_label: str
    hint: str


class AmortizationActionResponse(BaseModel):
    action: str
    fixed_expense_item_id: int | None = None
    label: str
    hint: str | None = None


class DuplicateSuggestionResponse(BaseModel):
    plaid_transaction_id: int
    manual_transaction_id: int
    score: float
    confidence_label: str
    plaid_transaction: TransactionResponse
    manual_transaction: TransactionResponse


class TransactionListResponse(BaseModel):
    items: list[TransactionResponse]
    total: int
    page: int
    per_page: int
    categories: list[CategoryResponse]
    accounts: list[AccountResponse]
    duplicate_suggestions: list[DuplicateSuggestionResponse]
    budget_actions: dict[int, TransactionBudgetActionResponse] = Field(default_factory=dict)
    amortization_actions: dict[int, AmortizationActionResponse] = Field(default_factory=dict)
    # Flask 0ddefb0: page rows already linked to a recurring template.
    recurring_transaction_ids: list[int] = Field(default_factory=list)


class TransactionCreateRequest(BaseModel):
    posted_date: date
    description: str
    amount: float
    merchant: str | None = None
    account_id: int | None = None
    account_name: str | None = None
    category_id: int | None = None
    category_name: str | None = None
    notes: str | None = None


class TransactionCategoryUpdateRequest(BaseModel):
    category_id: int | None = None
    new_category_name: str | None = None
    apply_to_similar: bool = False
    # Flask 0ddefb0 mark_recurring: create/refresh a recurring expense
    # template from this transaction.
    mark_recurring: bool = False
    recurring_name: str | None = None
    recurring_start_date: str | None = None
    recurring_second_date: str | None = None
    recurring_frequency: str = "monthly"
    recurring_days_of_week: list[int | str] = Field(default_factory=list)
    recurring_monthly_week_numbers: list[int | str] = Field(default_factory=list)
    recurring_monthly_weekday: int | str | None = None


class TransactionCategoryUpdateResponse(BaseModel):
    # Mirrors the Flask update_transaction_category JSON payload (420b456
    # apply-to-similar + 964c369 budget activation).
    transaction: TransactionResponse
    updated_transaction_ids: list[int] = Field(default_factory=list)
    similar_updated_count: int = 0
    rule_created: bool = False
    created_budget_target: float | None = None
    budget_action: TransactionBudgetActionResponse | None = None
    amortization_action: AmortizationActionResponse | None = None
    recurring_message: str | None = None
    recurring_success: bool | None = None


class TransactionSplitInput(BaseModel):
    category_id: int
    amount: float
    notes: str | None = None


class TransactionSplitsUpdateRequest(BaseModel):
    clear_splits: bool = False
    splits: list[TransactionSplitInput] = Field(default_factory=list)


class DuplicateTransactionMergeRequest(BaseModel):
    first_transaction_id: int
    second_transaction_id: int


class DuplicateTransactionMergeResponse(BaseModel):
    merged: bool
    surviving_transaction: TransactionResponse
    deleted_transaction_id: int


class TransactionImportMapping(BaseModel):
    date: str
    description: str
    amount: str | None = None
    debit: str | None = None
    credit: str | None = None
    account: str | None = None


class TransactionImportPreviewRequest(BaseModel):
    csv_text: str | None = None
    csv_base64: str | None = None
    mapping: TransactionImportMapping | None = None
    fallback_account: str | None = None


class TransactionImportRow(BaseModel):
    posted_date: date
    description: str
    amount: float
    transaction_type: str
    source_name: str
    category_id: int | None = None
    category_name: str | None = None


class TransactionImportPreviewResponse(BaseModel):
    staged_import_id: str
    headers: list[str]
    sample_rows: list[dict]
    mapping: dict[str, str | None]
    new_transactions: list[TransactionImportRow]
    new_count: int
    duplicate_count: int


class TransactionImportStagedResponse(BaseModel):
    staged_import_id: str
    new_transactions: list[TransactionImportRow]
    new_count: int


class TransactionImportCommitRequest(BaseModel):
    confirm: bool = True


class TransactionImportCommitResponse(BaseModel):
    imported: int
    duplicate_count: int
    transactions: list[TransactionResponse]


class CategoryCreateRequest(BaseModel):
    name: str
    kind: str = "expense"
    # API adaptation of Flask create_category's return_to sniffing: when the
    # category is created from the transactions screen, Flask seeds an initial
    # budget target. The stateless API carries that intent explicitly.
    activate_budget: bool = False


class CategoryUpdateRequest(BaseModel):
    name: str | None = None
    kind: str | None = None


class CategoryDeleteRequest(BaseModel):
    replacement_category_id: int | None = None


class CategoryDeleteResponse(BaseModel):
    deleted_category_id: int
    replacement_category: CategoryResponse | None = None


class CategoryRuleCondition(BaseModel):
    field: str
    operator: str = "contains"
    value: str
    value_secondary: str = ""
    group: str = "primary"
    join: str = "and"


class CategoryRuleCreateRequest(BaseModel):
    category_id: int
    conditions: list[CategoryRuleCondition] = Field(default_factory=list, max_length=4)
    match_text: str | None = None


class CategoryRuleUpdateRequest(BaseModel):
    category_id: int
    conditions: list[CategoryRuleCondition] = Field(default_factory=list, max_length=4)
    match_text: str | None = None


class CategoryRuleDeleteRequest(BaseModel):
    confirm: bool = True


class CategoryRuleResponse(BaseModel):
    id: int
    category: CategoryResponse
    match_text: str
    match_type: str
    rule_logic: str
    conditions: list[dict]
    summary: str
    created_at: datetime
    updated_at: datetime
    applied_count: int | None = None


class CategoryRuleListResponse(BaseModel):
    rules: list[CategoryRuleResponse]
    categories: list[CategoryResponse]

File: apps/web/lib/transactions.ts
import type { components } from "@clearpath/api-client";
import { transactionImportPreviewViewSchema, transactionReviewViewSchema } from "@clearpath/validation";

type ApiMe = components["schemas"]["MeResponse"];
type ApiList = components["schemas"]["TransactionListResponse"];
type ApiTransaction = components["schemas"]["TransactionResponse"];
type ApiCategory = components["schemas"]["CategoryResponse"];
type ApiAccount = components["schemas"]["AccountResponse"];
type ApiPlaid = components["schemas"]["PlaidItemListResponse"];
type ApiPreview = components["schemas"]["TransactionImportPreviewResponse"];

function mapSession(data: ApiMe) {
  return {
    ownerUserId: data.id,
    householdName: data.household_name ?? null,
    selectedPlan: data.selected_plan,
    billingStatus: data.billing_status,
    planDisplayName: data.plan_display_name,
    primaryAccountHolder: data.primary_account_holder,
    subject: {
      id: data.session_subject.id,
      subjectType: data.session_subject.subject_type,
      email: data.session_subject.email,
      displayName: data.session_subject.display_name,
      firstName: data.session_subject.first_name,
      avatarInitial: data.session_subject.avatar_initial,
      householdRole: data.session_subject.household_role ?? null,
    },
    featureAccess: (data.feature_access ?? []).map((row) => ({
      feature: row.feature,
      enabled: row.enabled,
      hidden: row.hidden,
      requiredPlan: row.required_plan,
    })),
  };
}

function mapCategory(row: ApiCategory) {
  return {
    id: row.id,
    name: row.name,
    kind: row.kind,
    monthlyTarget: row.monthly_target,
    isDefault: row.is_default,
    budgetGroupKey: row.budget_group_key ?? null,
    budgetSortOrder: row.budget_sort_order ?? null,
    canManage: row.can_manage ?? false,
  };
}

function mapAccount(row: ApiAccount) {
  return {
    id: row.id,
    name: row.name,
    accountType: row.account_type,
    institution: row.institution ?? null,
    currentBalance: row.current_balance,
    isManual: row.is_manual,
    mask: row.mask ?? null,
  };
}

export function mapTransaction(row: ApiTransaction) {
  return {
    id: row.id,
    postedDate: row.posted_date,
    description: row.description,
    displayMerchant: row.display_merchant,
    rawDescription: row.raw_description ?? null,
    amount: row.amount,
    transactionType: row.transaction_type,
    sourceName: row.source_name ?? null,
    notes: row.notes ?? null,
    plaidTransactionId: row.plaid_transaction_id ?? null,
    plaidCategoryLabel: row.plaid_category_label ?? null,
    paymentChannelLabel: row.payment_channel_label ?? null,
    locationSummary: row.location_summary ?? null,
    pending: row.pending,
    category: row.category ? mapCategory(row.category) : null,
    account: row.account ? mapAccount(row.account) : null,
    splits: (row.splits ?? []).map((split) => ({
      id: split.id,
      category: mapCategory(split.category),
      amount: split.amount,
      notes: split.notes ?? null,
    })),
  };
}

export function mapTransactionReview(data: ApiList, me: ApiMe, plaid: ApiPlaid) {
  return transactionReviewViewSchema.safeParse({
    session: mapSession(me),
    items: data.items.map(mapTransaction),
    total: data.total,
    page: data.page,
    perPage: data.per_page,
    categories: data.categories.map(mapCategory),
    accounts: data.accounts.map(mapAccount),
    duplicateSuggestions: data.duplicate_suggestions.map((row) => ({
      plaidTransactionId: row.plaid_transaction_id,
      manualTransactionId: row.manual_transaction_id,
      score: row.score,
      confidenceLabel: row.confidence_label,
      plaidTransaction: mapTransaction(row.plaid_transaction),
      manualTransaction: mapTransaction(row.manual_transaction),
    })),
    budgetActions: Object.fromEntries(Object.entries(data.budget_actions ?? {}).map(([id, row]) => [id, {
      categoryName: row.category_name,
      target: row.target,
      targetLabel: row.target_label,
      hint: row.hint,
    }])),
    amortizationActions: Object.fromEntries(Object.entries(data.amortization_actions ?? {}).map(([id, row]) => [id, {
      action: row.action,
      fixedExpenseItemId: row.fixed_expense_item_id ?? null,
      label: row.label,
      hint: row.hint ?? null,
    }])),
    recurringTransactionIds: data.recurring_transaction_ids ?? [],
    plaidItems: plaid.items.map((item) => ({
      id: item.id,
      institutionName: item.institution_name ?? null,
      status: item.status,
      lastSyncedAt: item.last_synced_at ?? null,
      errorMessage: item.error_message ?? null,
      reconnectRequiredAt: item.reconnect_required_at ?? null,
      accounts: (item.accounts ?? []).map(mapAccount),
    })),
  });
}

export function mapTransactionImportPreview(data: ApiPreview) {
  return transactionImportPreviewViewSchema.safeParse({
    stagedImportId: data.staged_import_id,
    headers: data.headers,
    sampleRows: data.sample_rows,
    mapping: {
      date: data.mapping.date ?? "",
      description: data.mapping.description ?? "",
      amount: data.mapping.amount ?? null,
      debit: data.mapping.debit ?? null,
      credit: data.mapping.credit ?? null,
      account: data.mapping.account ?? null,
    },
    newTransactions: data.new_transactions.map((row) => ({
      postedDate: row.posted_date,
      description: row.description,
      amount: row.amount,
      transactionType: row.transaction_type,
      sourceName: row.source_name,
      categoryId: row.category_id ?? null,
      categoryName: row.category_name ?? null,
    })),
    newCount: data.new_count,
    duplicateCount: data.duplicate_count,
  });
}
